### PR TITLE
expression: fix tidb can't alter table from other-type with null value to timestamp with NOT NULL attribute

### LINKED ISSUE #29328: Error when Column definition is modified (ALTER...MODIFY)from NULL to NOT NULL
## Bug Report

Please answer these questions before submitting your issue. Thanks!
```sql
set sql_mode=default;
drop table if exists t1;
create table t1 (a bigint);
insert into t1 values (NULL);
alter table t1 modify a timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ;
```

### 2. What did you expect to see? (Required)
```sql
Query 1 of 5, Rows read: 0, Elapsed time (seconds) - Total: 0, SQL query: 0, Reading results: 0

Query 2 of 5, Rows read: 0, Elapsed time (seconds) - Total: 0.005, SQL query: 0.005, Reading results: 0

Query 3 of 5, Rows read: 0, Elapsed time (seconds) - Total: 0.005, SQL query: 0.005, Reading results: 0
1 Row(s) Inserted
Query 4 of 5, Rows read: 0, Elapsed time (seconds) - Total: 0.001, SQL query: 0.001, Reading results: 0

Query 5 of 5, Rows read: 0, Elapsed time (seconds) - Total: 0.012, SQL query: 0.012, Reading results: 0
```
### 3. What did you see instead (Required)
```sql
Query 1 of 5, Rows read: 0, Elapsed time (seconds) - Total: 0.001, SQL query: 0.001, Reading results: 0

Query 2 of 5, Rows read: 0, Elapsed time (seconds) - Total: 0.272, SQL query: 0.272, Reading results: 0

Query 3 of 5, Rows read: 0, Elapsed time (seconds) - Total: 0.08, SQL query: 0.08, Reading results: 0
1 Row(s) Inserted
Query 4 of 5, Rows read: 0, Elapsed time (seconds) - Total: 0.007, SQL query: 0.007, Reading results: 0
Error: Data truncated for column 'a' at row 1
SQLState:  01000
ErrorCode: 1265
Error occurred in:
alter table t1 modify a timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP
```
### 4. What is your TiDB version? (Required)
```sql
+-------------------------+--------------------------------------------------------------------------+
| Variable_name           | Value                                                                    |
+-------------------------+--------------------------------------------------------------------------+
| innodb_version          | 5.6.25                                                                   |
| protocol_version        | 10                                                                       |
| tidb_analyze_version    | 2                                                                        |
| tidb_row_format_version | 2                                                                        |
| tls_version             | TLSv1,TLSv1.1,TLSv1.2                                                    |
| version                 | 5.7.25-TiDB-v5.2.1                                                       |
| version_comment         | TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible |
| version_compile_machine | x86_64                                                                   |
| version_compile_os      | osx10.8                                                                  |
+-------------------------+--------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: ailinkid <314806019@qq.com>

### What problem does this PR solve?

Issue Number: close #29328

### What is changed and how it works?
After testing from mysql8, we found that there is a special case for timestamp to handle with null values. 
It will convert null value to current timestamp although the target timestamp type has a NOT NULL value.

**check the origin issue comments for more details**

### Check List

Tests 

- [x] Unit test
- [x] Integration test

### Release note

```release-note
None
```

### CODE REVIEW

File: table/column.go [hunk 1]
```diff
@@ -283,6 +283,12 @@ func handleZeroDatetime(ctx sessionctx.Context, col *model.ColumnInfo, casted ty
 // TODO: change the third arg to TypeField. Not pass ColumnInfo.
 func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, returnErr, forceIgnoreTruncate bool) (casted types.Datum, err error) {
 	sc := ctx.GetSessionVars().StmtCtx
+	if val.Kind() == types.KindNull && col.FieldType.Tp == mysql.TypeTimestamp {
+		if v, err := expression.GetTimeValue(ctx, strings.ToUpper(ast.CurrentTimestamp), col.Tp, int8(col.Decimal)); err == nil {
```
Comment (tangenta) [inline table/column.go:287]: Can you encapsulate the code about getting current timestamp into a function and reuse it directly?
```go
if upperX == strings.ToUpper(ast.CurrentTimestamp) {
    defaultTime, err := getStmtTimestamp(ctx)
    if err != nil {
        return d, err
    }
    value.SetCoreTime(types.FromGoTime(defaultTime.Truncate(time.Duration(math.Pow10(9-int(fsp))) * time.Nanosecond)))
    if tp == mysql.TypeTimestamp || tp == mysql.TypeDatetime {
        err = value.ConvertTimeZone(time.Local, ctx.GetSessionVars().Location())
        if err != nil {
            return d, err
        }
    }
}
```

Comment (AilinKid) [reply table/column.go]: make sense, done

File: ddl/column_type_change_test.go [hunk 2]
```diff
@@ -2227,3 +2227,50 @@ func (s *testColumnTypeChangeSuite) TestChangeFromBitToStringInvalidUtf8ErrMsg(c
...
+	// SOME dml cases.
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a timestamp NOT NULL)")
+	_, err = tk.Exec("insert into t values()")
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "[table:1364]Field 'a' doesn't have a default value")
+
+	_, err = tk.Exec("insert into t values(null)")
+	c.Assert(err, IsNil)
+	tk.MustQuery("select count(*) from t where a = null").Check(testkit.Rows("0"))
```
Comment (tangenta) [inline ddl/column_type_change_test.go:2275]:
```sql
mysql> create table t(a timestamp NOT NULL);
Query OK, 0 rows affected (0.04 sec)

mysql> insert into t values (null);
ERROR 1048 (23000): Column 'a' cannot be null
```

MySQL's result is different from the one in TiDB.

Comment (AilinKid) [reply ddl/column_type_change_test.go:2275]: it's controlled by `explicit_defaults_for_timestamp` as illustrated in the issue. I'll fix it

File: expression/helper.go [hunk 3]
```diff
@@ -56,26 +56,45 @@ func IsValidCurrentTimestampExpr(exprNode ast.ExprNode, fieldType *types.FieldTy
 	return (containsArg && isConsistent) || (!containsArg && !containsFsp)
 }
 
+// ExportedGetTimeCurrentTimestamp is used for generating a timestamp for some special cases: cast null value to timestamp type with not null flag.
+func ExportedGetTimeCurrentTimestamp(ctx sessionctx.Context, tp byte, fsp int8) (d types.Datum, err error) {
```
Comment (tangenta) [inline expression/helper.go:60]:
```suggestion
func GetTimeCurrentTimestamp(ctx sessionctx.Context, tp byte, fsp int8) (d types.Datum, err error) {
```

We don't have to add this prefix to show it is exported :)

File: table/column.go [hunk 4]
```diff
@@ -278,6 +278,12 @@ func handleZeroDatetime(ctx sessionctx.Context, col *model.ColumnInfo, casted ty
 // TODO: change the third arg to TypeField. Not pass ColumnInfo.
 func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, returnErr, forceIgnoreTruncate bool) (casted types.Datum, err error) {
 	sc := ctx.GetSessionVars().StmtCtx
+	if val.Kind() == types.KindNull && col.FieldType.Tp == mysql.TypeTimestamp && ctx.GetSessionVars().StmtCtx.IsDDLJobInQueue {
```
Comment (zimulala) [inline table/column.go:281]: Could we put this logic out of `CastValue` and before it in `getRowRecord`? Then we needn't affect other function who calls `CastValue`.

File: expression/helper.go [hunk 5]
```diff
@@ -56,26 +56,45 @@ func IsValidCurrentTimestampExpr(exprNode ast.ExprNode, fieldType *types.FieldTy
 	return (containsArg && isConsistent) || (!containsArg && !containsFsp)
 }
 
+// ExportedGetTimeCurrentTimestamp is used for generating a timestamp for some special cases: cast null value to timestamp type with not null flag.
+func GetTimeCurrentTimestamp(ctx sessionctx.Context, tp byte, fsp int8) (d types.Datum, err error) {
```
Comment (zimulala) [inline expression/helper.go:60]: `GetTimeCurrentTimestamp` isn't be used. And you use `ExportedGetTimeCurrentTimestamp` instead of this.

Comment (zimulala) [reply expression/helper.go:60]: Please add unit tests for this function.

File: expression/helper.go [hunk 6]
```diff
@@ -56,26 +56,45 @@ func IsValidCurrentTimestampExpr(exprNode ast.ExprNode, fieldType *types.FieldTy
...
+
+func getTimeCurrentTimeStamp(ctx sessionctx.Context, tp byte, fsp int8) (t types.Time, err error) {
+	value := types.NewTime(types.ZeroCoreTime, tp, fsp)
+	defaultTime, err := getStmtTimestamp(ctx)
+	if err != nil {
+		return value, err
+	}
+	value.SetCoreTime(types.FromGoTime(defaultTime.Truncate(time.Duration(math.Pow10(9-int(fsp))) * time.Nanosecond)))
+	if tp == mysql.TypeTimestamp || tp == mysql.TypeDatetime {
+		err = value.ConvertTimeZone(time.Local, ctx.GetSessionVars().Location())
```
Comment (zimulala) [inline expression/helper.go:78]: Why here use local not session's timezone?

File: ddl/column_type_change_test.go [hunk 7]
```diff
@@ -2238,3 +2238,49 @@ func (s *testColumnTypeChangeSuite) TestForIssue24621(c *C) {
 	errMsg := "[types:1265]Data truncated for column 'a', value is '0123456789abc'"
 	tk.MustGetErrMsg("alter table t modify a char(12) null;", errMsg)
 }
+
+func (s *testColumnTypeChangeSuite) TestChangeNullValueFromOtherTypeToTimestamp(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+
+	// SOME ddl cases.
```
Comment (zimulala) [inline ddl/column_type_change_test.go:2246]:
```suggestion
	// Some ddl cases.
```

File: ddl/column_type_change_test.go [hunk 8]
```diff
@@ -2238,3 +2238,49 @@ func (s *testColumnTypeChangeSuite) TestForIssue24621(c *C) {
...
+		tk.MustQuery("select * from t").Check(testkit.Rows("<nil>"))
+	}
+
+	prepare2()
+	// only from other type NULL to timestamp type NOT NULL, it should be successful. (timestamp to timestamp excluded)
+	_, err = tk.Exec("alter table t modify column a timestamp NOT NULL")
+	c.Assert(err, NotNil)
+	c.Assert(err.Error(), Equals, "[ddl:1265]Data truncated for column 'a' at row 1")
+
+	// SOME dml cases.
```
Comment (zimulala) [inline ddl/column_type_change_test.go:2277]:
```suggestion
	// Some dml cases.
```

**Review by zimulala — APPROVED** [review]
LGTM

**Review by tangenta — APPROVED** [review]

**Pull request merged**